File: src/can/can_error.cpp
#include "can/can_error.h"

#include <google/protobuf/util/json_util.h>
#include <tuple>

namespace Serializers
{
namespace Can
{
Invalid_can_id::Invalid_can_id(const PbCan::Invalid_can_id& protobuf) {
    timestamp = protobuf.timestamp();
    id = protobuf.id();
    payload = {protobuf.payload().begin(), protobuf.payload().end()};
    network = protobuf.network();
}

Invalid_can_id::operator PbCan::Invalid_can_id() const {
    PbCan::Invalid_can_id ret;
    ret.set_timestamp(timestamp);
    ret.set_id(id);
    *(ret.mutable_payload()) = {payload.begin(), payload.end()};
    ret.set_network(network);
    return ret;
}

std::string Invalid_can_id::serializeAsJsonString() const {
    PbCan::Invalid_can_id protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string Invalid_can_id::serializeAsProtobufString() const {
    PbCan::Invalid_can_id protobuf(*this);
    return protobuf.SerializeAsString();
}

bool Invalid_can_id::deserializeFromJsonString(const std::string& str) {
    PbCan::Invalid_can_id protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool Invalid_can_id::deserializeFromProtobufString(const std::string& str) {
    PbCan::Invalid_can_id protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}
}
}

File: py/can/can_error.py
from dataclasses import dataclass, field
from enum import Enum
from typing import List, Dict

from can import can_error_pb2
from google.protobuf.json_format import MessageToJson, Parse


@dataclass
class Invalid_can_id:
    timestamp: int = 0
    id: int = 0
    payload: List[int] = field(default_factory=list)
    network: str = ""
    
    _proto_message: can_error_pb2.Invalid_can_id = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = can_error_pb2.Invalid_can_id()

    def _populate_proto(self):
        self._proto_message.timestamp = self.timestamp
        self._proto_message.id = self.id
        del self._proto_message.payload[:]
        for val in self.payload:
            self._proto_message.payload.append(val)
        self._proto_message.network = self.network

    @classmethod
    def _from_proto(cls, proto_message) -> "Invalid_can_id":
        return cls(
            timestamp = proto_message.timestamp,
            id = proto_message.id,
            payload=[int(val) for val in proto_message.payload],
            network = proto_message.network,
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "Invalid_can_id":
        message = can_error_pb2.Invalid_can_id()
        message.ParseFromString(data)
        return cls(
            timestamp = message.timestamp,
            id = message.id,
            payload = [int(val) for val in message.payload],
            network = message.network,
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "Invalid_can_id":
        message = can_error_pb2.Invalid_can_id()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())
